sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuExpressions.scala

```diff
@@ -278,47 +312,27 @@ trait GpuTernaryExpression extends TernaryExpression with GpuExpression {
-              withResource(GpuScalar.from(v2, child2.dataType)) { scalar2 =>
-                GpuColumnVector.from(doColumnar(v0, v1, scalar2), dataType)
-              }
-            case (v0, v1, v2: GpuColumnVector) =>
-              withResource(GpuScalar.from(v0, child0.dataType)) { scalar0 =>
-                withResource(GpuScalar.from(v1, child1.dataType)) { scalar1 =>
-                  GpuColumnVector.from(doColumnar(scalar0, scalar1, v2), dataType)
-                }
-              }
-            case (v0, v1: GpuColumnVector, v2) =>
-              withResource(GpuScalar.from(v0, child0.dataType)) { scalar0 =>
-                withResource(GpuScalar.from(v2, child2.dataType)) { scalar2 =>
-                  GpuColumnVector.from(doColumnar(scalar0, v1, scalar2), dataType)
-                }
-              }
-            case (v0: GpuColumnVector, v1, v2) =>
-              withResource(GpuScalar.from(v1, child1.dataType)) { scalar1 =>
-                withResource(GpuScalar.from(v2, child2.dataType)) { scalar2 =>
-                  GpuColumnVector.from(doColumnar(v0, scalar1, scalar2), dataType)
-                }
-              }
-            case (v0, v1, v2) if v0 != null && v1 != null && v2 != null =>
-              withResource(GpuScalar.from(v0, child0.dataType)) { v0Scalar =>
-                withResource(GpuScalar.from(v1, child1.dataType)) { v1Scalar =>
-                  withResource(GpuScalar.from(v2, child2.dataType)) { v2Scalar =>
-                    GpuColumnVector.from(doColumnar(batch.numRows(), v0Scalar, v1Scalar, v2Scalar),
-                      dataType)
-                  }
-                }
-              }
-            case _ => null
+              GpuColumnVector.from(doColumnar(v0, v1, v2), dataType)
+            case (v0: GpuScalar, v1: GpuColumnVector, v2: GpuColumnVector) =>
+              GpuColumnVector.from(doColumnar(v0, v1, v2), dataType)
+            case (v0: GpuColumnVector, v1: GpuScalar, v2: GpuColumnVector) =>
+              GpuColumnVector.from(doColumnar(v0, v1, v2), dataType)
+            case (v0: GpuColumnVector, v1: GpuColumnVector, v2: GpuScalar) =>
+              GpuColumnVector.from(doColumnar(v0, v1, v2), dataType)
+            case (v0: GpuScalar, v1: GpuScalar, v2: GpuColumnVector) =>
+              GpuColumnVector.from(doColumnar(v0, v1, v2), dataType)
+            case (v0: GpuScalar, v1: GpuColumnVector, v2: GpuScalar) =>
+              GpuColumnVector.from(doColumnar(v0, v1, v2), dataType)
+            case (v0: GpuColumnVector, v1: GpuScalar, v2: GpuScalar) =>
+              GpuColumnVector.from(doColumnar(v0, v1, v2), dataType)
+            case (v0: GpuScalar, v1: GpuScalar, v2: GpuScalar) =>
+              GpuColumnVector.from(doColumnar(batch.numRows(), v0, v1, v2), dataType)
+            // null is not welcome, return a null column instead.
+            case (v0, v1, v2) if v0 == null || v1 == null || v2 == null =>
```

Comment:
This existed before so it is not the end of the world, but I would like a follow on issue for us to look at and understand why this code assumes a null safe eval for GpuTernaryExpression when the CPU code does not.

Reply:
Yes this existed before in both of the `GpuBinaryExpression ` and `GpuTernaryExpression`. Just keep it here since I do not know the history/background.
Here is the followup issue https://github.com/NVIDIA/spark-rapids/issues/2412